drivers/ethernet/eth_xilinx_axi_ethernet_lite.c

```diff
@@ -0,0 +1,609 @@
+		length_addr = XILINX_AXI_ETHERNET_LITE_TX_PONG_LENGTH_REG_OFFSET;
+		control_addr = XILINX_AXI_ETHERNET_LITE_TX_PONG_CTRL_REG_OFFSET;
+	} else {
+		buffer_addr = XILINX_AXI_ETHERNET_LITE_TX_PING_START_REG_OFFSET;
+		length_addr = XILINX_AXI_ETHERNET_LITE_TX_PING_LENGTH_REG_OFFSET;
+		control_addr = XILINX_AXI_ETHERNET_LITE_TX_PING_CTRL_REG_OFFSET;
+	}
+
+	if (config->has_interrupt) {
+		(void)k_sem_take(&data->tx_sem, K_FOREVER);
+	}
+
+	if (xilinx_axi_ethernet_lite_read_reg(config, control_addr) &
+	    XILINX_AXI_ETHERNET_LITE_TX_PING_CTRL_BUSY_MASK) {
+		/*
+		 * no interrupt -> try to transmit as many packets as the L2 wants, discard them if
+		 * busy; otherwise, semaphore for flow control
+		 */
+		if (config->has_interrupt) {
+			LOG_WRN("Unexpectedly, %s buffer is busy!",
+				control_addr == XILINX_AXI_ETHERNET_LITE_TX_PING_CTRL_REG_OFFSET
+					? "ping"
+					: "pong");
+		}
+
+		net_pkt_unref(pkt);
+		return -EBUSY;
+	}
+
+	data->tx_ping_toggle = !data->tx_ping_toggle;
+
+	/* no need to linearize - can copy fragments one by one into transmit buffer */
+	do {
+		int frag_len = cursor->buf->len;
+		const uint8_t *frag_data = cursor->buf->data;
+
+		xilinx_axi_ethernet_lite_write_transmit_buffer(config, buffer_addr, frag_data,
+							       frag_len);
+
+		buffer_addr += frag_len;
+	} while (xilinx_axi_ethernet_lite_cursor_advance(cursor));
+
+	xilinx_axi_ethernet_lite_write_reg(config, length_addr, net_pkt_get_len(pkt));
+
+	/* as API is asynchronous, need not wait for TX completion */
+	xilinx_axi_ethernet_lite_write_reg(config, control_addr,
+					   XILINX_AXI_ETHERNET_LITE_TX_PING_TX_MASK);
+
+	return 0;
+}
+
+static const struct ethernet_api xilinx_axi_ethernet_lite_api = {
+	.get_phy = xilinx_axi_ethernet_lite_get_phy,
+	.get_capabilities = xilinx_axi_ethernet_lite_get_caps,
+	.iface_api.init = xilinx_axi_ethernet_lite_iface_init,
+	.set_config = xilinx_axi_ethernet_lite_set_config,
+	.send = xilinx_axi_ethernet_lite_send};
+
+static inline int
+xilinx_axi_ethernet_lite_read_to_pkt(const struct xilinx_axi_ethernet_lite_config *config,
+				     struct net_pkt *pkt, mem_addr_t buffer_addr,
+				     size_t bytes_to_read)
+{
+	int ret = 0;
+
+	for (size_t read_bytes = 0; read_bytes < bytes_to_read; read_bytes += sizeof(uint32_t)) {
+		uint32_t current_data = xilinx_axi_ethernet_lite_read_reg(config, buffer_addr);
+		size_t bytes_to_write_now = read_bytes + sizeof(uint32_t) > bytes_to_read
+						    ? bytes_to_read - read_bytes
+						    : sizeof(uint32_t);
+
+		ret += net_pkt_write(pkt, &current_data, bytes_to_write_now);
+
+		if (ret < 0) {
+			LOG_ERR("Write error bytes %zu/%zu (%zu)", read_bytes, bytes_to_read,
+				bytes_to_write_now);
+		} else {
+			LOG_DBG("Write OK bytes %zu/%zu (%zu) cursor %p remaining %d", read_bytes,
+				bytes_to_read, bytes_to_write_now, pkt->cursor.buf,
+				pkt->cursor.buf ? pkt->cursor.buf->size - pkt->cursor.buf->len : 0);
+		}
+
+		buffer_addr += sizeof(uint32_t);
+	}
+	return ret;
+}
```

Comment:
just read it at one. take a look at https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_litex_liteeth.c for reference